addons/sourcemod/scripting/gokz-core/misc.sp

```diff
@@ -605,6 +605,176 @@ void OnMapStart_FixMissingSpawns()
 	}
 }
 
+// =====[ BUTTONS ]=====
+
+void OnClientPreThinkPost_UseButtons(int client)
+{
+	if (GOKZ_GetCoreOption(client, Option_ButtonThroughPlayers) == ButtonThroughPlayers_Enabled && GetEntProp(client, Prop_Data, "m_afButtonPressed") & IN_USE)
+	{
+		int entity = FindUseEntity(client);
+		if (entity != -1)
+		{
+			AcceptEntityInput(entity, "Use", client, client, 1);
+		}
+	}
+}
+
+static int FindUseEntity(int client)
+{
+	float fwd[3];
+	float angles[3];
+	GetClientEyeAngles(client, angles);
+	GetAngleVectors(angles, fwd, NULL_VECTOR, NULL_VECTOR);
+
+	Handle trace;
+
+	float eyeOrigin[3];
+	GetClientEyePosition(client, eyeOrigin);
+	int useableContents = (MASK_NPCSOLID_BRUSHONLY | MASK_OPAQUE_AND_NPCS) & ~CONTENTS_OPAQUE;
+
+	float endpos[3];
+
+	// Check if +use trace collide with a player first, so we don't activate any button twice
+	trace = TR_TraceRayFilterEx(eyeOrigin, angles, useableContents, RayType_Infinite, TRFOtherPlayersOnly, client);
+	if (TR_DidHit(trace))
+	{
+		int ent = TR_GetEntityIndex(trace);
+		if (ent < 1 || ent > MaxClients)
+		{
+			return -1;
+		}
+		// Search for a button behind it.
+		trace = TR_TraceRayFilterEx(eyeOrigin, angles, useableContents, RayType_Infinite, TraceEntityFilterPlayers);
+		if (TR_DidHit(trace))
+		{
+			char buffer[20];
+			ent = TR_GetEntityIndex(trace);
+			// Make sure that it is a button, and this button activates when pressed.
+			// If it is not a button, check its parent to see if it is a button.
+			bool isButton;
+			while (ent != -1)
+			{
+				GetEntityClassname(ent, buffer, sizeof(buffer));
+				if (StrEqual("func_button", buffer, false) && GetEntProp(ent, Prop_Data, "m_spawnflags") & SF_BUTTON_USE_ACTIVATES)
+				{
+					isButton = true;
+					break;
+				}
+				else
+				{
+					ent = GetEntPropEnt(ent, Prop_Data, "m_hMoveParent");
+				}
+			}
+			if (isButton)
+			{
+				TR_GetEndPosition(endpos, trace);
+				float delta[3];
+				for (int i = 0; i < 2; i++)
+				{
+					delta[i] = endpos[i] - eyeOrigin[i];
+				}
+				// Z distance is treated differently.
+				float m_vecMins[3];
+				float m_vecMaxs[3];
+				float m_vecOrigin[3];
+				GetEntPropVector(ent, Prop_Send, "m_vecOrigin", m_vecOrigin);
+				GetEntPropVector(ent, Prop_Send, "m_vecMins", m_vecMins);
+				GetEntPropVector(ent, Prop_Send, "m_vecMaxs", m_vecMaxs);
+
+				delta[2] = IntervalDistance(endpos[2], m_vecOrigin[2] + m_vecMins[2], m_vecOrigin[2] + m_vecMaxs[2]);
+				if (GetVectorLength(delta) < 80.0)
+				{
+					return ent;
+				}
+			}
+		}
+	}
+	
```

Comment:
What's the reason for using the sphere method as well, shouldn't the code above be enough?

Reply:
The sphere check allows you to press the button without looking directly at it (see kz_11342 where you can press the start timer even when looking away from it), it also extends the player's reach by an extra 2 units to 82u. It is usually the sphere check that allows for the most optimal timer tech setup, actually.